File: js/ClTableau.js
load('CL.js')
load('ClLinearExpression.js')
load('ClVariable.js')


var ClTableau = new Class({
  /* FIELDS:
      var _columns //Hashtable
      var _rows //Hashtable
      var _infeasibleRows //Set
      var _externalRows //Set
      var _externalParametricVars //Set
 */
  initialize: function() {
    this._columns = {};
    this._rows = {};
    this._infeasibleRows = {}/* SET: */;
    this._externalRows = {}/* SET: */;
    this._externalParametricVars = {}/* SET: */;
  },
  noteRemovedVariable: function(v /*ClAbstractVariable*/, subject /*ClAbstractVariable*/) {
    if (this.fTraceOn) this.fnenterprint("noteRemovedVariable: " + v + ", " + subject);
    if (subject != null) {
      delete (/* Set */this._columns[v])[subject];
    }
  },
  noteAddedVariable: function(v /*ClAbstractVariable*/, subject /*ClAbstractVariable*/) {
    if (this.fTraceOn) this.fnenterprint("noteAddedVariable: " + v + ", " + subject);
    if (subject != null) {
      this.insertColVar(v, subject);
    }
  },
  getInternalInfo: function() {
    var retstr = "Tableau Information:\n";
    retstr += "Rows: " + Object.keys(this._rows).length;
    retstr += " (= " + (Object.keys(this._rows).length - 1) + " constraints)";
    retstr += "\nColumns: " + Object.keys(this._columns).length;
    retstr += "\nInfeasible Rows: " + Object.keys(this._infeasibleRows).length;
    retstr += "\nExternal basic variables: " + Object.keys(this._externalRows).length;
    retstr += "\nExternal parametric variables: ";
    retstr += Object.keys(this._externalParametricVars).length;
    retstr += "\n";
    return retstr;
  },
  toString: function() {
    var bstr = "Tableau:\n";
    for (clv in this._rows) {
      var expr = /* ClLinearExpression */this._rows[clv];
      bstr += clv;
      bstr += " <==> ";
      bstr += expr;
      bstr += "\n";
    }
    bstr += "\nColumns:\n";
    bstr += CL.hashToString(this._columns);
    bstr += "\nInfeasible rows: ";
    bstr += CL.setToString(this._infeasibleRows);
    bstr += "External basic variables: ";
    bstr += CL.setToString(this._externalRows);
    bstr += "External parametric variables: ";
    bstr += CL.setToString(this._externalParametricVars);
    return bstr;
  },
  insertColVar: function(param_var /*ClAbstractVariable*/, rowvar /*ClAbstractVariable*/) {
    var rowset = /* Set */this._columns[param_var];
    if (rowset == null) this._columns[param_var] = rowset = {}/* SET: */;
    rowset.insert(rowvar);
  },
  addRow: function(aVar /*ClAbstractVariable*/, expr /*ClLinearExpression*/) {
    if (this.fTraceOn) this.fnenterprint("addRow: " + aVar + ", " + expr);
    this._rows[aVar] = expr;
    for (clv in expr) {
      this.insertColVar(clv, aVar);
      if (clv.isExternal()) {
        this._externalParametricVars[clv] = true;
      }
    }
    if (aVar.isExternal()) {
      this._externalRows[aVar] = true;
    }
    if (this.fTraceOn) this.traceprint(this.toString());
  },
  removeColumn: function(aVar /*ClAbstractVariable*/) {
    if (this.fTraceOn) this.fnenterprint("removeColumn:" + aVar);
    var rows = /* Set */delete this._columns[aVar];
    if (rows != null) {
      for (clv in rows) {
        var expr = /* ClLinearExpression */this._rows[clv];
        expr.terms().remove(aVar);
      }
    }
    else {
      if (this.fTraceOn) this.debugprint("Could not find var " + aVar + " in _columns");
    }
    if (aVar.isExternal()) {
      delete this._externalRows[aVar];
      delete this._externalParametricVars[aVar];
    }
  },
  removeRow: function(aVar /*ClAbstractVariable*/) {
    if (this.fTraceOn) this.fnenterprint("removeRow:" + aVar);
    var expr = /* ClLinearExpression */this._rows[aVar];
    this.assert(expr != null);
    for (clv in expr) {
      var varset = /* Set */this._columns[clv];
      if (varset != null) {
        if (this.fTraceOn) this.debugprint("removing from varset " + aVar);
        varset.remove(aVar);
      }
    }
    delete this._infeasibleRows[aVar];
    if (aVar.isExternal()) {
      delete this._externalRows[aVar];
    }
    delete this._rows[aVar];
    if (this.fTraceOn) this.fnexitprint("returning " + expr);
    return expr;
  },
  substituteOut: function(oldVar /*ClAbstractVariable*/, expr /*ClLinearExpression*/) {
    if (this.fTraceOn) this.fnenterprint("substituteOut:" + oldVar + ", " + expr);
    if (this.fTraceOn) this.traceprint(this.toString());
    var varset = /* Set */this._columns[oldVar];
    for (v in varset) {
      var row = /* ClLinearExpression */this._rows[v];
      row.substituteOut(oldVar, expr, v, this);
      if (v.isRestricted() && row.constant() < 0.0) {
        this._infeasibleRows[v] = true;
      }
    }
    if (oldVar.isExternal()) {
      this._externalRows[oldVar] = true;
      delete this._externalParametricVars[oldVar];
    }
    delete this._columns[oldVar];
  },
  columns: function() {
    return this._columns;
  },
  rows: function() {
    return this._rows;
  },
  columnsHasKey: function(subject /*ClAbstractVariable*/) {
    return (this._columns[subject] != undefined);
  },
  rowExpression: function(v /*ClAbstractVariable*/) {
    return /* ClLinearExpression */this._rows[v];
  },
});

File: java/ClTableau.java
// $Id$
//
// Cassowary Incremental Constraint Solver
// Original Smalltalk Implementation by Alan Borning
// This Java Implementation by Greg J. Badros, <gjb@cs.washington.edu>
// http://www.cs.washington.edu/homes/gjb
// (C) 1998, All Rights Reserved.
//
// (c) 1998 Alan Borning and Greg Badros.  This code is provided for use by
// students for course projects in the course CSE 595 in the Department of
// Computer Science and Engineering, University of Washington, during winter
// quarter 1998.  Any other use requires written permission from the copyright
// holders.
//
// ClTableau

import java.util.*;

class ClTableau
{

  public void noteRemovedVariable(ClAbstractVariable v, ClAbstractVariable subject)
    { 
      //#ifndef CL_NO_TRACE
      //Tracer TRACER(__FUNCTION__);
      /** @c2j++ Replacement from cerr << "(" << v << ", " << subject << ")" << endl; */
      //System.err.println("(" + String.valueOf(v) + ", " + String.valueOf(subject) + ")");
      //#endif
      if (subject != null) {
	((Set) my_columns.get(v)).remove(subject);
      }
    }

  public void noteAddedVariable(ClAbstractVariable v, ClAbstractVariable subject)
    { 
      //#ifndef CL_NO_TRACE
      //Tracer TRACER(__FUNCTION__);
      // System.err.println("(" + String.valueOf(v) + ", " + String.valueOf(subject) + ")");
      //#endif
      if (subject != null) {
	insertColVar(v,subject);
      }
    }
  
  public String toString()
    { 
      StringBuffer bstr = new StringBuffer("Tableau:\n");
      bstr.append(my_rows.toString());

      bstr.append("Columns:\n");
      bstr.append(my_columns.toString());
      
      bstr.append("Infeasible rows:\n");
      bstr.append(my_infeasibleRows.toString());

      bstr.append("External basic variables:\n");
      bstr.append(my_externalRows.toString());

      bstr.append("External parametric variables:\n");
      bstr.append(my_externalParametricVars.toString());
      
      return bstr.toString();
    }

  private void insertColVar(ClAbstractVariable param_var, 
			    ClAbstractVariable rowvar)
  { 
    Set rowset = (Set) my_columns.get(param_var);
    if (rowset == null)
      my_columns.put(param_var,rowset = new Set());
    rowset.insert(rowvar);
  }

  protected void addRow(ClAbstractVariable var, ClLinearExpression expr)
    {
      //#ifndef CL_NO_TRACE
      // Tracer TRACER(__FUNCTION__);
      // System.err.println("(" + String.valueOf(var) + ", " + String.valueOf(expr) + ")");
      //#endif

      // for each variable in expr, add var to the set of rows which
      // have that variable in their expression

      for (Enumeration e = expr.terms().keys() ; e.hasMoreElements(); ) {
        ClVariable clv = (ClVariable) e.nextElement();
	insertColVar(clv,var);
      }
      if (var.isExternal()) {
	my_externalRows.insert(var);
      }
      //#ifndef CL_NO_TRACE
      //System.err.println(String.valueOf(*this));
      //#endif
    }

  protected void removeColumn(ClAbstractVariable var)
    {
      //#ifndef CL_NO_TRACE
      //Tracer TRACER(__FUNCTION__);
      //System.err.println("(" + String.valueOf(var) + ")");
      //#endif
      // remove the rows with the variables in varset

      Set rows = (Set) my_columns.remove(var);

      for (Enumeration e = rows.elements() ; e.hasMoreElements(); ) {
        ClVariable clv = (ClVariable) e.nextElement();
	ClLinearExpression expr = (ClLinearExpression) my_rows.get(clv);
	expr.terms().remove(clv);
      }
      
      if (var.isExternal()) {
	my_externalRows.remove(var);
	my_externalParametricVars.remove(var);
      }
    }

  protected ClLinearExpression removeRow(ClAbstractVariable var)
    {
      //#ifndef CL_NO_TRACE
      //  Tracer TRACER(__FUNCTION__);
      //System.err.println("(" + var.toString() + ")");
      //#endif

      ClLinearExpression expr = (ClLinearExpression) my_rows.remove(var);

      for (Enumeration e = expr.terms().keys() ; e.hasMoreElements(); ) {
        ClVariable clv = (ClVariable) e.nextElement();
	((Set)my_columns.get(clv)).remove(var);
	my_infeasibleRows.remove(var);
      }

      if (var.isExternal())
	{
	my_externalRows.remove(var);
	}
      //#ifndef CL_NO_TRACE
      //System.err.println("- returning " + expr.toString());
      //#endif
      return expr;
    }

  protected void substituteOut(ClAbstractVariable oldVar, ClLinearExpression expr)
    {
      //#ifndef CL_NO_TRACE
      //System.err.print("* ClTableau::");
      //Tracer TRACER(__FUNCTION__);
      //System.err.println("(" + String.valueOf(oldVar) + ", " + String.valueOf(expr) + ")");
      //System.err.println(String.valueOf((*this)));
      //#endif
      
      Set varset = (Set) my_columns.get(oldVar);
      for (Enumeration e = varset.elements(); e.hasMoreElements(); ) {
	ClAbstractVariable v = (ClAbstractVariable) e.nextElement();
	ClLinearExpression row = (ClLinearExpression) my_rows.get(v);
	row.substituteOut(oldVar,expr,v,this);
	if (v.isRestricted() && row.constant() < 0.0) {
	  my_infeasibleRows.insert(v);
	}
      }

      if (oldVar.isExternal()) {
	my_externalRows.insert(oldVar);
	my_externalParametricVars.remove(oldVar);
      }
      my_columns.remove(oldVar);
    }

  protected Hashtable columns()
    { return my_columns; }

  protected Hashtable rows()
    { return my_rows; }

  protected boolean columnsHasKey(ClAbstractVariable subject)
    { 
      return my_columns.containsKey(subject);
    }

  protected ClLinearExpression rowExpression(ClAbstractVariable v)
    {
      return (ClLinearExpression) my_rows.get(v);
    }

  private Hashtable my_columns; // From ClAbstractVariable to Set of variables
  private Hashtable my_rows;    // From ClAbstractVariable to ClLinearExpression
  private Set my_infeasibleRows; // Set of ClAbstractVariable-s
  private Set my_externalRows; // Set of ClVariable-s
  private Set my_externalParametricVars; // Set of ClVariable-s

}
